### PR TITLE
[HPRO-1318] Group samples/orders based upon when they're made.

### PR DESCRIPTION
**Pull request by Ehaic**
| Q                   | A
| ------------------- | --------------
| Target Release?     | 3.0.0 
| Bug fix?            | ❌ 
| New feature?        | ❌ 
| Database migration? | ✅ 
| New configuration?  | ❌ 
| Composer updates?   | ❌ 
| NPM updates?        | ❌ 
| Jira Ticket(s)      | HPRO-1318

### CODE REVIEW

**Comment by shyamzcode** [comment]
@Ehaic Could you also fix phpstan errors?

**Comment by shyamzcode** [comment]
@Ehaic Noticed in the participant summary page it's displaying multiple orders for the same timepoint blood samples.

image

**Comment by Ehaic** [comment]
> @Ehaic Noticed in the participant summary page it's displaying duplicate orders for the same timepoint blood samples.
> 
> image

Good catch, I changed `templates/program/nph/order/samples.html.twig` because I found another bug testing these changes that made it so those row's only printed out if the first sample in a particular timepoint was created. I'll make more changes to that loop to fix it.

File: migrations/Version20230118144015.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public function up(Schema $schema) : void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+        $this->addSql('ALTER TABLE nph_samples add column sample_group VARCHAR(100) not null');
+    }
+
+    public function down(Schema $schema) : void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
Comment (thornview) [inline migrations/Version20230118144015.php:29]: Another comment to remove. 
```suggestion
```

File: migrations/Version20230118144015.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+declare(strict_types=1);
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (thornview) [inline migrations/Version20230118144015.php:12]: Remove auto-generated comments

```suggestion
```

File: src/Controller/NphOrderController.php [hunk 3]
```diff
@@ -71,7 +71,7 @@ public function generateOrderAction(
             'stoolSamples' => $nphOrderService->getSamplesByType('stool'),
             'nailSamples' => $nphOrderService->getSamplesByType('nail'),
             'samplesOrderIds' => $nphOrderService->getSamplesWithOrderIds(),
-            'samplesStatus' => $nphOrderService->getSamplesWithStatus()
+            'samplesStatus' => $nphOrderService->getSamplesWithStatus(),
```
Comment (thornview) [inline src/Controller/NphOrderController.php:74]: Unnecessary trailing comma
```suggestion
            'samplesStatus' => $nphOrderService->getSamplesWithStatus()
```

File: src/Controller/NphPDFController.php [hunk 4]
```diff
@@ -26,4 +26,20 @@ public function render_pdf(
...
+    public function renderPDFSampleGroup(
+        $participantId,
+        $module,
+        $visit,
+        $sampleGroup,
+        PDFService $PDF,
+        NphOrderService $nphOrderService,
+        NphParticipantSummaryService $nphParticipantSummaryService
+    ): Response {
+        $OrderPDF = $PDF->batchPDF($nphOrderService->getParticipantOrderSummaryByModuleVisitAndSampleGroup($participantId, $module, $visit, $sampleGroup)['order'], $nphParticipantSummaryService->getParticipantById($participantId), $module, $visit);
```
Comment (thornview) [inline src/Controller/NphPDFController.php:42]: It's hard to read the logic when the parameters span multiple lines.  Break this into separate lines to make it easier to read. 

```suggestion
        $OrderPDF = $PDF->batchPDF($nphOrderService->
            getParticipantOrderSummaryByModuleVisitAndSampleGroup(
                $participantId, 
                $module, 
                $visit,
                $sampleGroup)['order'],
                $nphParticipantSummaryService->getParticipantById($participantId), 
                $module, 
                $visit
        );
```

File: src/Entity/NphOrder.php [hunk 5]
```diff
@@ -268,4 +268,14 @@ public function canModify($type): bool
...
     }
+
+    public function getSampleGroupBySampleCode($sampleCode): string
+    {
+        foreach ($this->nphSamples as $nphSample) {
+            if ($nphSample->getSampleCode() === $sampleCode) {
+                return $nphSample->getSampleGroup();
+            }
+        }
+        return '';
```
Comment (thornview) [inline src/Entity/NphOrder.php:279]: It seems odd to return an empty string.  Would this work if the function returned null?

File: src/Repository/NphOrderRepository.php [hunk 6]
```diff
@@ -30,4 +30,17 @@ public function getOrdersByVisitType($participantId, $visitType): array
             ->getQuery()
             ->getResult();
     }
+
+    public function getOrdersBySampleGroup($participantId, $sampleGroup)
```
Comment (thornview) [inline src/Repository/NphOrderRepository.php:34]: Type-hint these parameters and provide a return type.  Also, this new public method should have a related test.

File: src/Service/Nph/NphOrderService.php [hunk 7]
```diff
@@ -399,6 +421,18 @@ public function getParticipantOrderSummaryByModuleAndVisit(string $participantid
         return $orderSummary;
     }
 
+    public function getParticipantOrderSummaryByModuleVisitAndSampleGroup(
```
Comment (thornview) [inline src/Service/Nph/NphOrderService.php:424]: New public method should have a test.

File: src/Service/Nph/NphOrderService.php [hunk 8]
```diff
@@ -192,6 +193,24 @@ public function generateSampleId(): string
...
+                $id = null;
+            } else {
+                break;
+            }
+        }
+        if (empty($id)) {
+            throw new \Exception('Failed to generate unique sample group');
+        }
+        return $id;
+    }
```
Comment (thornview) [inline src/Service/Nph/NphOrderService.php:212]: This method should also have tests, which will probably need some mocks.

**Review by thornview — CHANGES_REQUESTED** [review]

**Review by shyamzcode — APPROVED** [review]

**Pull request merged**